---
title: "STING Pathway in Sjogren's Disease, supplementary figures for bulk code"
author: "Joseph Twohig"
output:
  html_document:
    toc: true
    toc-location: left
  pdf_document:
    toc: true
---

```{r message=FALSE, warning=FALSE, results = FALSE}
library(readxl)
library(tidyverse) 
library(edgeR) # DGE testing
library(DESeq2) # DGE testing
library(clusterProfiler) # GO
library(AnnotationDbi) # GO
library(patchwork) # Plotting
library(gridExtra) # Plotting
library(ggrepel) # Plotting
library(GSVA) # Performing GSVA
library(msigdbr) # Pulling gene sets
library(ComplexHeatmap) # For heatmaps
library(org.Hs.eg.db) # Get gene mapping
library(circlize)  # For color functions
library(dendextend)  # For dendrogram manipulation
library(scales) # Color scales
library(ggpubr) # Stat tests
library(WGCNA)
library(CorLevelPlot)

setwd("/data/Warner_Lab/bulk_sting_analysis")
```

Load in expression matrices and metadata

```{r}
metadata <- read.csv("metadata.csv")
metadata$SSA <- case_when(
  grepl("1", metadata$SSA) ~ "Positive",
  grepl("0", metadata$SSA) ~ "Negative",
  TRUE ~ NA_character_
)
metadata$SSB <- case_when(
  grepl("1", metadata$SSB) ~ "Positive",
  grepl("0", metadata$SSB) ~ "Negative",
  TRUE ~ NA_character_
)

gland <- read.delim("/data/Warner_Lab/STING_BulkRNA_SjD_HV_nonSjD/MSG_with_corrected_sample_names/report/RawCountFile_filtered.txt",row.names=1,check.names=FALSE)%>% 
  dplyr::select(metadata$MSG_Samples)

blood <- read.delim("/data/Warner_Lab/STING_BulkRNA_SjD_HV_nonSjD/Blood_with_corrected_sample_names/report/RawCountFile_filtered.txt",row.names=1,check.names=FALSE) %>%
  dplyr::select(metadata$Blood_Samples)
```

Convert the Ensembl IDs in the counts matrices to gene symbols

```{r message=FALSE, warning=FALSE}
# Make function to convert and filter for gene symbols
ens_to_symbol <- function(matrix) {
  ens <- rownames(matrix)
  symbols <- mapIds(org.Hs.eg.db, keys = ens,
                    column = c('SYMBOL'), keytype = 'ENSEMBL')
  symbols <- symbols[!is.na(symbols)]
  symbols <- symbols[!duplicated(symbols)]
  symbols <- symbols[match(rownames(matrix), names(symbols))]
  symbols <- symbols[!is.na(symbols)]
  matrix <- matrix[names(symbols), ]
  rownames(matrix) <- symbols
  return(matrix)
}

# Convert gland and blood count matrices
gland <- ens_to_symbol(gland)
blood <- ens_to_symbol(blood)
```

# Supplement 1a: Extended Gland Gene Ontology (GO) analysis

```{r}
condition <- factor(metadata$disease[match(colnames(gland), metadata$MSG_Samples)])
coldata <- data.frame(row.names = colnames(gland), condition)

# Create DESeq2 dataset
dds <- DESeqDataSetFromMatrix(countData = gland, colData = coldata, design = ~condition)
dds <- DESeq(dds)

# Get differential expression results (SjD vs nonSjD/HV)
res <- results(dds, contrast = c("condition", "SjD", "nonSjD/HV"))

# Filter significant DEGs
sigs <- na.omit(res)
sigs <- sigs[sigs$padj < 0.05 & sigs$baseMean > 50, ]

# Select upregulated genes (log2FC > 0.5)
genes_to_test <- rownames(sigs[sigs$log2FoldChange > 0.5, ])

# Perform GO enrichment analysis (Biological Process)
GO_results <- enrichGO(gene = genes_to_test, OrgDb = org.Hs.eg.db, keyType = "SYMBOL", ont = "BP")

# Generate barplot
plot <- barplot(GO_results, showCategory = 16, main = "GO Enrichment Blood SjD vs nonSjD/HV")
plot

# ggsave(file="supplement/glandGObarplotextended.svg", plot=plot, width=8, height=16)
```

# Supplement 1b: Blood Gene Ontology (GO) analysis

```{r message=FALSE, warning=FALSE}
# Reformat metadata to make single comparison
metadata$disease <- case_when(
  grepl("nonSjD", metadata$condition) ~ "nonSjD/HV",
  grepl("HV", metadata$condition) ~ "nonSjD/HV",
  TRUE ~ "SjD"
)

condition <- factor(metadata$disease[match(colnames(blood), metadata$Blood_Samples)])
coldata <- data.frame(row.names = colnames(blood), condition)

# Create DESeq2 dataset
dds <- DESeqDataSetFromMatrix(countData = blood, colData = coldata, design = ~condition)
dds <- DESeq(dds)

# Get differential expression results (SjD vs nonSjD/HV)
res <- results(dds, contrast = c("condition", "SjD", "nonSjD/HV"))

# Filter significant DEGs
sigs <- na.omit(res)
sigs <- sigs[sigs$padj < 0.05 & sigs$baseMean > 50, ]

# Select upregulated genes (log2FC > 0.5)
genes_to_test <- rownames(sigs[sigs$log2FoldChange > 0.5, ])

# Perform GO enrichment analysis (Biological Process)
GO_results <- enrichGO(gene = genes_to_test, OrgDb = org.Hs.eg.db, keyType = "SYMBOL", ont = "BP")

# Generate barplot
plot <- barplot(GO_results, showCategory = 8, main = "GO Enrichment Blood SjD vs nonSjD/HV")
plot

# ggsave(file="supplement/bloodGObarplot.svg", plot=plot, width=8, height=8)
```

# Supplement 1c: Gene Set Variation Analysis (GSVA), blood heatmap

```{r message=FALSE, warning=FALSE}
# Retrieve pathway data from msigdbr, filtering relevant pathways
pathways <- rbind(msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG_LEGACY") %>% 
  filter(gs_name %in% c("KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY", "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY")))

# Create gene set for GSVA input
gene_sets <- split(pathways,
                   pathways$gs_name) %>% 
  lapply(`[[`,
         "gene_symbol")
gene_sets$IFN_Type_I_28genes <- c("CXCL10","DDX60","EPSTI1","GBP1","HERC5",
                        "HERC6","IFI27","IFI44","IFI44L","IFI6",
                        "IFIT1","IFIT2","IFIT3","IFIT5","ISG15",
                        "LAMP3","LY6E","MX1","OAS1","OAS2",
                        "OAS3","OASL","RSAD2","RTP4","SIGLEC1",
                        "SOCS1","SPATS2L","USP18")
gene_sets$IFN_Type_II_8genes <- c("FCGR1A","CXCL9","IFI30","CIITA","SERPING1",
                        "IRF1","CXCL10","GBP8")

# Normalize counts data matrices before running GSVA
norm_blood <- cpm(blood, log = T)
gsva_blood <- gsva(gsvaParam(exprData = norm_blood, geneSets = gene_sets))
blood_heat <- gsva_blood
rownames(blood_heat) <- c("DNA Sensing", "RNA Sensing","28 gene Type I IFN", "8 gene Type II IFN")

# Creating a dataframe with GSVA results and metadata to use in visualization
targets <- metadata %>% 
  dplyr::select("SPIT_number", "condition")

# Heatmap annotation
col_condition <- c("SjD" = "red", "nonSjD" = "grey", "HV" = "blue")

# Construct heatmap
bloodGSVAHeatmap <- Heatmap(
  blood_heat,
  column_title = "Blood GSVA Heatmap",
  name = "Expression",
  show_row_names = TRUE,
  show_column_names = FALSE,
  row_names_gp = gpar(fontsize = 10),
  column_names_gp = gpar(fontsize = 10),
  heatmap_legend_param = list(
    title_gp = gpar(fontsize = 12),
    at = seq(-2, 2, length.out = 5)
  ),
  cluster_rows = TRUE,
  cluster_columns = TRUE,
  top_annotation = HeatmapAnnotation(
    Condition = targets$condition,
    col = list(
      Condition = col_condition
    ),
    annotation_legend_param = list(
      title_gp = gpar(fontsize = 12)
    )
  )
)
bloodGSVAHeatmap

svg(file="supplement/bloodGSVAHeatmap.svg", width=8, height=5.5)
draw(bloodGSVAHeatmap)
dev.off()
png(file="supplement/bloodGSVAHeatmap.png", width=7, height=5.5, units = "in", res = 500)
draw(bloodGSVAHeatmap)
dev.off()
```

# Supplement 2a: Expression dot plot for Blood

```{r message=FALSE, warning=FALSE}
goiDotPlotWithSignificance <- function(matrix, goi_list, tissue_type) {
  if (tissue_type == "gland") {
    var_name = "MSG_Samples"
    plot_title = "Minor Salivary Gland"
  } else {
    var_name = "Blood_Samples"
    plot_title = "Blood"
  }
  
  # Normalize gene expression (CPM)
  goi_mat <- as.data.frame(cpm(matrix))
  goi_mat <- goi_mat %>% 
    filter(row.names(goi_mat) %in% goi_list) %>% 
    t() %>% 
    as.data.frame() %>% 
    rownames_to_column(var = var_name)
  
  # Merge with metadata
  goi_mat <- merge(goi_mat, metadata, by = var_name)
  
  # Reshape to long format
  goi_long <- goi_mat %>%
    pivot_longer(cols = all_of(goi_list), names_to = "gene", values_to = "expression")
  
  # Add z-scores
  goi_long <- goi_long %>%
    group_by(gene) %>%
    mutate(z_score = scale(expression)) %>%
    ungroup()
  
  # Order genes
  goi_long$gene <- factor(goi_long$gene, levels = goi_list)
  
  # Compute % expressed and avg z-score
  summary_data <- goi_long %>%
    mutate(is_expressed = expression > 0) %>%
    group_by(condition, gene) %>%
    summarize(
      percent_expressed = 100 * mean(is_expressed),
      average_expression = mean(z_score, na.rm = TRUE),
      .groups = "drop"
    )

  # ==== Significance Testing ====
  # Wilcoxon test for each gene
  stat_results <- goi_long %>%
    filter(condition %in% c("SjD", "HV")) %>%
    group_by(gene) %>%
    summarise(
      p_value = wilcox.test(expression ~ condition)$p.value,
      .groups = "drop"
    ) %>%
    mutate(significance = case_when(
      p_value < 0.001 ~ "***",
      p_value < 0.01 ~ "**",
      p_value < 0.05 ~ "*",
      TRUE ~ ""
    ))

  # Merge for plotting
  summary_data <- left_join(summary_data, stat_results, by = "gene")

  # Plot
  g <- ggplot(summary_data, aes(x = gene, y = condition)) +
    geom_point(aes(size = percent_expressed, color = average_expression)) +
    scale_size_continuous(name = "Percent Expression", range = c(1, 10)) +
    scale_color_gradient2(name = "Normalized Expression", low = "blue", mid = "white", high = "red", midpoint = 0) +
    geom_text(
      data = stat_results,
      aes(x = gene, y = 3.3, label = significance), # assuming only two groups: HV and SjD
      inherit.aes = FALSE,
      size = 5
    ) +
    theme_minimal() +
    labs(
      title = paste0(plot_title, " Normalized Gene Expression by Disease State"),
      x = "Gene", y = "Condition"
    ) +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1),
      axis.title = element_text(size = 12),
      plot.title = element_text(size = 14)
    )

  return(g)
}

# Genes of interest
goi <- c("STING1","CGAS", # Core genes
         "TBK1","IRF3","IKBKE", # Downstream signaling
         "STAT1","STAT2","IRF7", # Type I IFN response
         "MX1","ISG15","OAS1","OAS2","OAS3","IFIH1", # Feedback/Effectors
         "USP18","SOCS1","SOCS3", # Negative regulators
         "IL6","TNF","CXCL8","CXCL10" # Cytokines
         )

bloodgoi_sig <- goiDotPlotWithSignificance(matrix = blood, goi_list = goi, tissue_type = "blood")
bloodgoi_sig

# ggsave(file="supplement/bloodGoiDotplotSignificance.svg", plot=bloodgoi_sig, width=8.5, height=4.5)
```

# Supplement 2b-k: GSVA correlation plots

Prepare data frame for plotting

```{r message=FALSE, warning=FALSE}
# Retrieve pathway data from msigdbr, filtering relevant pathways
pathways <- rbind(msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG_LEGACY") %>% 
  filter(gs_name %in% c("KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY", "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY")))

# Create gene set for GSVA input
gene_sets <- split(pathways,
                   pathways$gs_name) %>% 
  lapply(`[[`,
         "gene_symbol")
gene_sets$IFN_Type_I_28genes <- c("CXCL10","DDX60","EPSTI1","GBP1","HERC5",
                        "HERC6","IFI27","IFI44","IFI44L","IFI6",
                        "IFIT1","IFIT2","IFIT3","IFIT5","ISG15",
                        "LAMP3","LY6E","MX1","OAS1","OAS2",
                        "OAS3","OASL","RSAD2","RTP4","SIGLEC1",
                        "SOCS1","SPATS2L","USP18")
gene_sets$IFN_Type_II_8genes <- c("FCGR1A","CXCL9","IFI30","CIITA","SERPING1",
                        "IRF1","CXCL10","GBP8")

# Normalize counts data matrices before running GSVA
norm_blood <- cpm(blood, log = T)
gsva_blood <- gsva(gsvaParam(exprData = norm_blood, geneSets = gene_sets))
gsva_blood_t <- as.data.frame(t(gsva_blood))
gsva_blood_t$Sample <- rownames(gsva_blood_t)

# Creating a dataframe with all GSVA results and metadata to use in visualization
targets <- metadata %>% 
  dplyr::select("SPIT_number","Blood_Samples","condition","SSA","Type.I.IFN.Bioassay.AVG") %>% 
  rename("Sample" = Blood_Samples)

merged_data <- merge(gsva_blood_t, targets, by = "Sample")

scatterplot <- function(df, x_var, y_var, col, shape_, x_title, y_title, col_title, shape_title = NULL, title_, facet.by, legend_pos) {
  plot <- ggplot(df %>%
                   filter(!is.na({{ y_var }})),
                 aes(x = {{ x_var }}, y = {{ y_var }}, color = {{ col }})) +
    geom_point(aes(shape = {{ shape_ }}), size = 3) + 
    geom_smooth(method = "lm", se = FALSE, formula = 'y ~ x') +  # Add a line of best fit
    labs(x = x_title,
         y = y_title,
         title = title_,
         color = col_title, 
         shape = shape_title) +
    scale_color_manual(values = c("SjD" = "red", "nonSjD" = "grey", "HV" = "blue")) +  # Set custom colors
    theme_minimal() +
    theme(legend.position = legend_pos) +
    theme(axis.text = element_text(size = 10),
          axis.title = element_text(size = 10),
          plot.title = element_text(size = 12)) +
    stat_cor(method = "pearson", size = 4, color = "black") +
    facet_wrap(vars({{ facet.by }}), scales = 'fixed')
  return(plot)
}
```

```{r}
norm_gland <- cpm(gland, log = T)
gsva_gland <- gsva(gsvaParam(exprData = norm_gland, geneSets = gene_sets))
gsva_gland_t <- as.data.frame(t(gsva_gland))
gsva_gland_t$Sample <- rownames(gsva_gland_t)

# Creating a dataframe with all GSVA results and metadata to use in visualization
targets <- metadata %>% 
  dplyr::select("SPIT_number","MSG_Samples","condition","SSA","Type.I.IFN.Bioassay.AVG") %>% 
  rename(Sample = "MSG_Samples")

merged_data <- merge(gsva_gland_t, targets, by = "Sample")


gland_dna_vs_type1 <- scatterplot(df = merged_data,
           x_var = KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY,
           y_var = IFN_Type_I_28genes,
           col = condition,
           # shape_ = SSA,
           x_title = "DNA Sensing Pathway Score",
           y_title = "Type I IFN 28 Gene Panel Score",
           title_ = "Gland: DNA sensing vs Type I IFN",
           col_title = NULL,
           # shape_title = "SSA",
           facet.by = condition,
           legend_pos = "right")

gland_dna_vs_type1
# ggsave(file="supplement/glandDNAvsT1.svg", plot=gland_dna_vs_type1, width=8, height=3)

gland_dna_vs_type2 <- scatterplot(df = merged_data,
           x_var = KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY,
           y_var = IFN_Type_II_8genes,
           col = condition,
           # shape_ = SSA,
           x_title = "DNA Sensing Pathway Score",
           y_title = "Type II IFN 8 Gene Panel Score",
           title_ = "Gland: DNA sensing vs Type II IFN",
           col_title = NULL,
           # shape_title = "SSA",
           facet.by = condition,
           legend_pos = "right")

gland_dna_vs_type2
# ggsave(file="supplement/glandDNAvsT2.svg", plot=gland_dna_vs_type2, width=8, height=3)

gland_dna_vs_bioassay <- scatterplot(df = merged_data,
           x_var = KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY,
           y_var = Type.I.IFN.Bioassay.AVG,
           col = condition,
           # shape_ = SSA,
           x_title = "DNA Sensing Pathway Score",
           y_title = "IFN Bioassay Score",
           title_ = "Gland: DNA sensing vs IFN Bioassay",
           col_title = NULL,
           # shape_title = "SSA",
           facet.by = condition,
           legend_pos = "right")

gland_dna_vs_bioassay
# ggsave(file="supplement/glandDNAvsIFNBioassay.svg", plot=gland_dna_vs_bioassay, width=8, height=3)

gland_rna_vs_type1 <- scatterplot(df = merged_data,
           x_var = KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY,
           y_var = IFN_Type_I_28genes,
           col = condition,
           # shape_ = SSA,
           x_title = "RNA Sensing Pathway Score",
           y_title = "Type I IFN 28 Gene Panel Score",
           title_ = "Gland: RNA sensing vs Type I IFN",
           col_title = NULL,
           # shape_title = "SSA",
           facet.by = condition,
           legend_pos = "right")

gland_rna_vs_type1
# ggsave(file="supplement/glandRNAvsT1.svg", plot=gland_rna_vs_type1, width=8, height=3)

gland_rna_vs_type2 <- scatterplot(df = merged_data,
           x_var = KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY,
           y_var = IFN_Type_II_8genes,
           col = condition,
           # shape_ = SSA,
           x_title = "RNA Sensing Pathway Score",
           y_title = "Type II IFN 8 Gene Panel Score",
           title_ = "Blood: RNA sensing vs Type II IFN",
           col_title = NULL,
           # shape_title = "SSA",
           facet.by = condition,
           legend_pos = "right")

gland_rna_vs_type2
# ggsave(file="supplement/glandRNAvsT2.svg", plot=gland_rna_vs_type2, width=8, height=3)
```

```{r}
blood_dna_vs_type1 <- scatterplot(df = merged_data,
           x_var = KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY,
           y_var = IFN_Type_I_28genes,
           col = condition,
           # shape_ = SSA,
           x_title = "DNA Sensing Pathway Score",
           y_title = "Type I IFN 28 Gene Panel Score",
           title_ = "Blood: DNA sensing vs Type I IFN",
           col_title = NULL,
           # shape_title = "SSA",
           facet.by = condition,
           legend_pos = "right")

blood_dna_vs_type1
# ggsave(file="supplement/bloodDNAvsT1.svg", plot=blood_dna_vs_type1, width=8, height=3)

blood_dna_vs_type2 <- scatterplot(df = merged_data,
           x_var = KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY,
           y_var = IFN_Type_II_8genes,
           col = condition,
           # shape_ = SSA,
           x_title = "DNA Sensing Pathway Score",
           y_title = "Type II IFN 8 Gene Panel Score",
           title_ = "Blood: DNA sensing vs Type II IFN",
           col_title = NULL,
           # shape_title = "SSA",
           facet.by = condition,
           legend_pos = "right")

blood_dna_vs_type2
# ggsave(file="supplement/bloodDNAvsT2.svg", plot=blood_dna_vs_type2, width=8, height=3)

blood_dna_vs_bioassay <- scatterplot(df = merged_data,
           x_var = KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY,
           y_var = Type.I.IFN.Bioassay.AVG,
           col = condition,
           # shape_ = SSA,
           x_title = "DNA Sensing Pathway Score",
           y_title = "IFN Bioassay Score",
           title_ = "Blood: DNA sensing vs IFN Bioassay",
           col_title = NULL,
           # shape_title = "SSA",
           facet.by = condition,
           legend_pos = "right")

blood_dna_vs_bioassay
# ggsave(file="supplement/bloodDNAvsIFNBioassay.svg", plot=blood_dna_vs_bioassay, width=8, height=3)

blood_rna_vs_type1 <- scatterplot(df = merged_data,
           x_var = KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY,
           y_var = IFN_Type_I_28genes,
           col = condition,
           # shape_ = SSA,
           x_title = "RNA Sensing Pathway Score",
           y_title = "Type I IFN 28 Gene Panel Score",
           title_ = "Blood: RNA sensing vs Type I IFN",
           col_title = NULL,
           # shape_title = "SSA",
           facet.by = condition,
           legend_pos = "right")

blood_rna_vs_type1
# ggsave(file="supplement/bloodRNAvsT1.svg", plot=blood_rna_vs_type1, width=8, height=3)

blood_rna_vs_type2 <- scatterplot(df = merged_data,
           x_var = KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY,
           y_var = IFN_Type_II_8genes,
           col = condition,
           # shape_ = SSA,
           x_title = "RNA Sensing Pathway Score",
           y_title = "Type II IFN 8 Gene Panel Score",
           title_ = "Blood: RNA sensing vs Type II IFN",
           col_title = NULL,
           # shape_title = "SSA",
           facet.by = condition,
           legend_pos = "right")

blood_rna_vs_type2
# ggsave(file="supplement/bloodRNAvsT2.svg", plot=blood_rna_vs_type2, width=8, height=3)
```

# Supplement 2l: Deconvolution proportions for gland

This data was generated from deconvolution.R

```{r}
metadata$SPIT_number <- as.character(metadata$SPIT_number)

# Format Gland
deconv_gland <- readRDS("deconv_celltypes_gland.RDS") %>%
  rownames_to_column(var = "Cell_Type") %>%
  pivot_longer(-Cell_Type, names_to = "SPIT_number", values_to = "Proportion") %>%
  left_join(metadata, by = "SPIT_number")

# Create color panels for plotting
cell_order_gland <- c(
  # Fibroblasts
  "Fibroblast",
  # Other Stromal
  "Pericyte","Endothelial",
  # Epithelial
  "Seromucous","Mucous","Duct","Ionocyte","Myoepithelial",
  # Plasma
  "IgG Plasma","IgA Plasma",
  # T cell
  "CD8 GZMK+ T","CD8 T","CD4 T","NK",
  #APC
  "B","Myeloid"
)
cell_colors_gland <- c(
  "Fibroblast" = "#1b6639",
  "Pericyte" = "#bd28d4",
  "Endothelial" = "#7b07f7",
  "Seromucous" = "#d92a0f",
  "Mucous" = "#d9960f",
  "Duct" = "#57241a",
  "Ionocyte" = "#fcc603",
  "Myoepithelial" = "#99000d",
  "IgA Plasma" = "#327a0d",
  "IgG Plasma" = "#94f50c",
  "CD8 GZMK+ T" = "#3182bd",
  "CD8 T" = "#6baed6",
  "CD4 T" = "#9ecae1",
  "NK" = "#c6dbef",
  "B" = "#1b04c7",
  "Myeloid" = "#29165c"
)
cell_colors_gland <- cell_colors_gland[cell_order_gland]
deconv_gland$Cell_Type <- factor(deconv_gland$Cell_Type, levels = cell_order_gland)

# Format blood
deconv_blood <- readRDS("deconv_celltypes_blood.RDS") %>%
  rownames_to_column(var = "Cell_Type") %>%
  pivot_longer(-Cell_Type, names_to = "SPIT_number", values_to = "Proportion") %>%
  left_join(metadata, by = "SPIT_number")

# Create color panels for plotting
cell_order_blood <- c(
  # T Cells
  "CD4 Naive","CD4 Tcm","CD4 Tem","Treg","CD8 Naive","CD8 Tcm","CD8 Tem","DNT",
  # B Cells
  "B","Plasmablast",
  #NK Cells
  "NK","Proliferating NK",
  # Myeloid Cells
  "CD14+ Monocyte","CD16+ Monocyte","mDC","pDC",
  # Progenitor/Stem
  "LPC","HSC"
)

cell_colors_blood <- c(
  "CD4 Naive" = "#fa2205",
  "CD4 Tcm" = "#a11a08",
  "CD4 Tem" = "#6b2103",
  "Treg" = "#edde0e",
  "CD8 Naive" = "#fa9405",
  "CD8 Tcm" = "#b88e11",
  "CD8 Tem" = "#786a02",
  "DNT" = "#140401",
  "B" = "#79dde8",
  "Plasmablast" = "#05fa98",
  "NK" = "#079107",
  "Proliferating NK" = "#05ff05",
  "CD14+ Monocyte" = "#0330fc",
  "CD16+ Monocyte" = "#03f4fc",
  "mDC" = "#0c034a",
  "pDC" = "#7703fc",
  "LPC" = "#ff0055",
  "HSC" = "#e833df"
)
cell_colors_blood <- cell_colors_blood[cell_order_blood]
deconv_blood$Cell_Type <- factor(deconv_blood$Cell_Type, levels = cell_order_blood)

# Gland
# Proportions
deconvGlandCondition <- ggplot(deconv_gland, aes(x = condition, y = Proportion, fill = Cell_Type)) +
  stat_summary(geom = "bar", fun = "mean", position = "stack") +
  theme_minimal() +
  scale_fill_manual(values = cell_colors_gland) +
  labs(title = "Gland: Deconvolution Cell Proportions by Condition",
       x = "Condition",
       y = "Mean Proportion",
       fill = "Cell Type") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
deconvGlandCondition
# ggsave("supplement/deconvGlandCondition.svg", deconvGlandCondition, width = 8, height = 6)
# Bar plots based on SSA status
deconvGlandSSA <- ggplot(deconv_gland, aes(x = SSA, y = Proportion, fill = Cell_Type)) +
  stat_summary(geom = "bar", fun = "mean", position = "stack") +
  theme_minimal() +
  scale_fill_manual(values = cell_colors_gland) +
  labs(title = "Gland: Deconvolution Cell Proportions by SSA Status",
       x = "SSA Status",
       y = "Mean Proportion",
       fill = "Cell Type") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
deconvGlandSSA
# ggsave("supplement/deconvGlandSSA.svg", deconvGlandSSA, width = 8, height = 6)

# Blood
# Proportions
deconvBloodCondition <- ggplot(deconv_blood, aes(x = condition, y = Proportion, fill = Cell_Type)) +
  stat_summary(geom = "bar", fun = "mean", position = "stack") +
  theme_minimal() +
  scale_fill_manual(values = cell_colors_blood) +
  labs(title = "Blood: Deconvolution Cell Proportions by Condition",
       x = "Condition",
       y = "Mean Proportion",
       fill = "Cell Type") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
deconvBloodCondition
# ggsave("supplement/deconvBloodCondition.svg", deconvBloodCondition, width = 8, height = 6)
# Bar plots based on SSA status
deconvBloodSSA <- ggplot(deconv_blood, aes(x = SSA, y = Proportion, fill = Cell_Type)) +
  stat_summary(geom = "bar", fun = "mean", position = "stack") +
  theme_minimal() +
  scale_fill_manual(values = cell_colors_blood) +
  labs(title = "Blood: Deconvolution Cell Proportions by SSA Status",
       x = "SSA Status",
       y = "Mean Proportion",
       fill = "Cell Type") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
deconvBloodSSA
# ggsave("supplement/deconvBloodSSA.svg", deconvBloodSSA, width = 8, height = 6)
```

# Supplement 3a,b: WGCNA

Gland

```{r message=FALSE, warning=FALSE}
# Prepare input for WGCNA
de_input = as.matrix(gland)
dds <- DESeqDataSetFromMatrix(round(de_input),
                              metadata,
                              design = ~condition)
dds <- DESeq(dds)
vsd <- varianceStabilizingTransformation(dds)
wpn_vsd <- getVarianceStabilizedData(dds)

# Filter the normalized expression matrix for top DEGs
expr_normalized <- wpn_vsd[rowVars(wpn_vsd) > quantile(rowVars(wpn_vsd), 0.95),] %>% 
  t()

allowWGCNAThreads(nThreads = 8) 
powers = c(c(1:10), seq(from = 12, to = 20, by = 2))
sft = pickSoftThreshold(
  expr_normalized,
  powerVector = powers,
  verbose = 5
)

expr_normalized[] <- sapply(expr_normalized, as.numeric)
soft_power <- 11
temp_cor <- cor
cor <- WGCNA::cor

# memory estimate w.r.t blocksize
bwnet <- blockwiseModules(expr_normalized,
                          maxBlockSize = 5000,
                          TOMType = "signed",
                          power = soft_power,
                          mergeCutHeight = 0.25,
                          numericLabels = FALSE,
                          randomSeed = 1234,
                          verbose = 3)
cor <- temp_cor

# Module Eigengenes
module_eigengenes <- bwnet$MEs
colnames(module_eigengenes) <- gsub("^ME", "", colnames(module_eigengenes))

# Number of genes for each module
table(bwnet$colors)

# See the clustering of genes
mergedColors <-  labels2colors(bwnet$colors)
plotDendroAndColors(
  bwnet$dendrograms[[1]],
  mergedColors[bwnet$blockGenes[[1]]],
  "Module colors",
  dendroLabels = FALSE,
  hang = 0.03,
  addGuide = TRUE,
  guideHang = 0.05 )

# Module-trait (eigengene-metadata) association
traits <- metadata %>% 
  dplyr::select("MSG_Samples", "Age", "Sex","condition","SSA","SSB","Focus.Score") %>% 
  mutate(Sex = ifelse(grepl('FEMALE', Sex), 1, 0)) %>% 
  mutate(HV = ifelse(grepl('HV', condition), 1, 0)) %>% 
  mutate(nonSjD = ifelse(grepl('nonSjD', condition), 1, 0)) %>% 
  mutate(SjD = startsWith(condition, "SjD")) %>% 
  mutate(SjD = ifelse(grepl('TRUE', SjD), 1, 0)) %>% 
  mutate(SSA = ifelse(grepl('Positive',SSA), 1, 0)) %>% 
  mutate(SSB = ifelse(grepl('Positive',SSB),1,0)) %>% 
  column_to_rownames(var="MSG_Samples")
traits$condition <- NULL

nSamples <- nrow(expr_normalized)
nGenes <- ncol(expr_normalized)

module.trait.corr <- cor(module_eigengenes, traits, use = 'p')
module.trait.corr.pvals <- corPvalueStudent(module.trait.corr, nSamples)

# Heatmap of module-trait assocaition
heatmap.data <- merge(module_eigengenes, traits, by = 'row.names')

heatmap.data <- heatmap.data %>% 
  column_to_rownames(var = 'Row.names')
WGCNA <- CorLevelPlot(heatmap.data,
             x = names(heatmap.data)[11:18],
             y = names(heatmap.data)[1:10],
             main = "Gland: Module-trait Relationships",
             col = c("blue1", "skyblue", "white", "pink", "red"))
WGCNA
svg(file="supplement/glandWGCNA.svg", width=12, height=8)
WGCNA
dev.off()
png(file="supplement/glandWGCNA.png", width=12, height=8, units = "in", res = 500)
WGCNA
dev.off()

# Compute module membership and p-values
geneModuleMembership <- as.data.frame(cor(t(norm_gland), module_eigengenes, use = "p"))
MMPvalue <- as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), 97))

# Get hub genes ranked by module membership (correlation)
glandWGCNAhublist <- list()
for (color in colnames(geneModuleMembership)) {
  # Merge p-values and correlations with gene symbols
  module_data <- merge(MMPvalue[color], geneModuleMembership[color], by = 0)
  colnames(module_data) <- c("Symbol", "p_value", "corr")
  
  # Filter genes assigned to this module
  module_genes <- as.data.frame(bwnet$colors) %>% filter(bwnet$colors == color)
  module_genes <- rownames(module_genes)
  # Filter, sort by membership, and select top genes
  top_genes <- module_data %>%
    filter(p_value < 0.01 & Symbol %in% module_genes & corr > 0) %>%
    arrange(desc(corr)) %>%
    pull(Symbol)
  
  glandWGCNAhublist[[color]] <- top_genes
}

# Find max gene count across all modules for padding
max_genes <- max(sapply(glandWGCNAhublist, length))

# Combine into a padded dataframe
glandWGCNAhubdf <- bind_rows(
  lapply(names(glandWGCNAhublist), function(color) {
    genes <- glandWGCNAhublist[[color]]
    padded_genes <- c(genes, rep(NA, max_genes - length(genes)))
    df <- as.data.frame(t(padded_genes), stringsAsFactors = FALSE)
    df$color <- color
    return(df)
  })
)

# Move color to first column and rename gene columns
glandWGCNAhubdf <- glandWGCNAhubdf %>%
  relocate(color) %>%
  setNames(c("color", paste0("Gene", seq_len(max_genes))))

# Save to CSV
write_csv(glandWGCNAhubdf, "supplement/glandWGCNAtop.csv")
```

Blood

```{r message=FALSE, warning=FALSE}
# Prepare input for WGCNA
de_input = as.matrix(blood)
dds <- DESeqDataSetFromMatrix(round(de_input),
                              metadata,
                              design = ~condition)
dds <- DESeq(dds)
vsd <- varianceStabilizingTransformation(dds)
wpn_vsd <- getVarianceStabilizedData(dds)

# Filter the normalized expression matrix for top DEGs
expr_normalized <- wpn_vsd[rowVars(wpn_vsd) > quantile(rowVars(wpn_vsd), 0.95),] %>% 
  t()

allowWGCNAThreads(nThreads = 8) 
powers = c(c(1:10), seq(from = 12, to = 20, by = 2))
sft = pickSoftThreshold(
  expr_normalized,
  powerVector = powers,
  verbose = 5
)

expr_normalized[] <- sapply(expr_normalized, as.numeric)
soft_power <- 11
temp_cor <- cor
cor <- WGCNA::cor

# memory estimate w.r.t blocksize
bwnet <- blockwiseModules(expr_normalized,
                          maxBlockSize = 5000,
                          TOMType = "signed",
                          power = soft_power,
                          mergeCutHeight = 0.25,
                          numericLabels = FALSE,
                          randomSeed = 1234,
                          verbose = 3)
cor <- temp_cor

# Module Eigengenes
module_eigengenes <- bwnet$MEs
colnames(module_eigengenes) <- gsub("^ME", "", colnames(module_eigengenes))

# Number of genes for each module
table(bwnet$colors)

# See the clustering of genes
mergedColors <-  labels2colors(bwnet$colors)
plotDendroAndColors(
  bwnet$dendrograms[[1]],
  mergedColors[bwnet$blockGenes[[1]]],
  "Module colors",
  dendroLabels = FALSE,
  hang = 0.03,
  addGuide = TRUE,
  guideHang = 0.05 )

# Module-trait (eigengene-metadata) association
traits <- metadata %>% 
  dplyr::select("Blood_Samples", "Age", "Sex","condition","SSA","SSB","Focus.Score") %>% 
  mutate(Sex = ifelse(grepl('FEMALE', Sex), 1, 0)) %>% 
  mutate(HV = ifelse(grepl('HV', condition), 1, 0)) %>% 
  mutate(nonSjD = ifelse(grepl('nonSjD', condition), 1, 0)) %>% 
  mutate(SjD = startsWith(condition, "SjD")) %>% 
  mutate(SjD = ifelse(grepl('TRUE', SjD), 1, 0)) %>% 
  mutate(SSA = ifelse(grepl('Positive',SSA), 1, 0)) %>% 
  mutate(SSB = ifelse(grepl('Positive',SSB),1,0)) %>% 
  column_to_rownames(var="Blood_Samples")
traits$condition <- NULL

nSamples <- nrow(expr_normalized)
nGenes <- ncol(expr_normalized)

module.trait.corr <- cor(module_eigengenes, traits, use = 'p')
module.trait.corr.pvals <- corPvalueStudent(module.trait.corr, nSamples)

# Heatmap of module-trait assocaition
heatmap.data <- merge(module_eigengenes, traits, by = 'row.names')

heatmap.data <- heatmap.data %>% 
  column_to_rownames(var = 'Row.names')
WGCNA <- CorLevelPlot(heatmap.data,
             x = names(heatmap.data)[12:19],
             y = names(heatmap.data)[1:11],
             main = "Blood: Module-trait Relationships",
             col = c("blue1", "skyblue", "white", "pink", "red"))
WGCNA
svg(file="supplement/bloodWGCNA.svg", width=12, height=8)
WGCNA
dev.off()
png(file="supplement/bloodWGCNA.png", width=12, height=8, units = "in", res = 500)
WGCNA
dev.off()

# Compute module membership and p-values
geneModuleMembership <- as.data.frame(cor(t(norm_blood), module_eigengenes, use = "p"))
MMPvalue <- as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), 97))

# Get hub genes ranked by module membership (correlation)
bloodWGCNAhublist <- list()
for (color in colnames(geneModuleMembership)) {
  # Merge p-values and correlations with gene symbols
  module_data <- merge(MMPvalue[color], geneModuleMembership[color], by = 0)
  colnames(module_data) <- c("Symbol", "p_value", "corr")
  
  # Filter genes assigned to this module
  module_genes <- as.data.frame(bwnet$colors) %>% filter(bwnet$colors == color)
  module_genes <- rownames(module_genes)
  # Filter, sort by membership, and select top genes
  top_genes <- module_data %>%
    filter(p_value < 0.01 & Symbol %in% module_genes & corr > 0) %>%
    arrange(desc(corr)) %>%
    pull(Symbol)
  
  bloodWGCNAhublist[[color]] <- top_genes
}

# Find max gene count across all modules for padding
max_genes <- max(sapply(bloodWGCNAhublist, length))

# Combine into a padded dataframe
bloodWGCNAhubdf <- bind_rows(
  lapply(names(bloodWGCNAhublist), function(color) {
    genes <- bloodWGCNAhublist[[color]]
    padded_genes <- c(genes, rep(NA, max_genes - length(genes)))
    df <- as.data.frame(t(padded_genes), stringsAsFactors = FALSE)
    df$color <- color
    return(df)
  })
)

# Move color to first column and rename gene columns
bloodWGCNAhubdf <- bloodWGCNAhubdf %>%
  relocate(color) %>%
  setNames(c("color", paste0("Gene", seq_len(max_genes))))

# Save to CSV
write_csv(bloodWGCNAhubdf, "supplement/bloodWGCNAtop.csv")
```
